src/DotNetWorker.Core/FunctionsApplication.cs

```diff
@@ -17,28 +19,52 @@ internal partial class FunctionsApplication : IFunctionsApplication
+        private readonly IFunctionInvocationDictionary _functionInvocationDictionary;
         private readonly IOptions<WorkerOptions> _workerOptions;
         private readonly ILogger<FunctionsApplication> _logger;
         private readonly IWorkerDiagnostics _diagnostics;
 
-        public FunctionsApplication(FunctionExecutionDelegate functionExecutionDelegate, IFunctionContextFactory functionContextFactory,
-             IOptions<WorkerOptions> workerOptions, ILogger<FunctionsApplication> logger, IWorkerDiagnostics diagnostics)
+        public FunctionsApplication(
+            FunctionExecutionDelegate functionExecutionDelegate,
+            IFunctionContextFactory functionContextFactory,
+            IFunctionInvocationDictionary functionInvocationDictionary,
+            IOptions<WorkerOptions> workerOptions,
+            ILogger<FunctionsApplication> logger,
+            IWorkerDiagnostics diagnostics)
         {
             _functionExecutionDelegate = functionExecutionDelegate ?? throw new ArgumentNullException(nameof(functionExecutionDelegate));
             _functionContextFactory = functionContextFactory ?? throw new ArgumentNullException(nameof(functionContextFactory));
+            _functionInvocationDictionary = functionInvocationDictionary ?? throw new ArgumentNullException(nameof(functionInvocationDictionary));
             _workerOptions = workerOptions ?? throw new ArgumentNullException(nameof(workerOptions));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _diagnostics = diagnostics ?? throw new ArgumentNullException(nameof(diagnostics));
         }
 
         public FunctionContext CreateContext(IInvocationFeatures features)
         {
+            CancellationTokenSource? cancellationTokenSource = null;
             var invocation = features.Get<FunctionInvocation>() ?? throw new InvalidOperationException($"The {nameof(FunctionInvocation)} feature is required.");
 
             var functionDefinition = _functionMap[invocation.FunctionId];
             features.Set<FunctionDefinition>(functionDefinition);
 
-            return _functionContextFactory.Create(features);
+            if (functionDefinition.IsCancellable)
+            {
+                cancellationTokenSource = new CancellationTokenSource();
+            }
+
+            var token = cancellationTokenSource is null ? CancellationToken.None : cancellationTokenSource.Token;
+            var context = _functionContextFactory.Create(features, token);
+
+            var invocationDetails = new FunctionInvocationDetails()
+                                        {
+                                            FunctionContext = context,
+                                            CancellationTokenSource = cancellationTokenSource,
+                                        };
+
+            _functionInvocationDictionary.TryAddInvocationDetails(context.InvocationId, invocationDetails);
```

Comment:
In addition to the disposal comment above, make sure disposal is also done if the `TryAddInvocationDetails` fails.